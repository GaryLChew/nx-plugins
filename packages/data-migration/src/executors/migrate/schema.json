{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "Migrate executor",
  "description": "",
  "type": "object",
  "properties": {
    "env": {
      "description": "The environment to use.",
      "type": "string",
      "default": "dev"
    },
    "cwd": {
      "description": "The directory where migrations are located.",
      "type": "string"
    },
    "migrationsDir": {
      "description": "The directory where migrations are located.",
      "type": "string",
      "default": "migrations"
    },
    "logLevel": {
      "description": "The log level.",
      "type": "string",
      "default": "info",
      "enum": ["error", "warn", "info", "debug"]
    },
    "lifecycleHook": {
      "description": "The lifecycle hook to run.",
      "type": "string",
      "default": "before:deploy",
      "enum": ["before:deploy", "after:deploy"]
    },
    "migrationTableName": {
      "description": "The name of the table to store the migration states.",
      "type": "string"
    }
  },
  "required": ["env", "cwd"]
}
