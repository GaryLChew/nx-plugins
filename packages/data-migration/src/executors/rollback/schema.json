{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "Rollback executor",
  "description": "",
  "type": "object",
  "properties": {
    "env": {
      "description": "The environment to use.",
      "type": "string",
      "default": "development"
    },
    "cwd": {
      "description": "The directory where migrations are located.",
      "type": "string"
    },
    "migrationsDir": {
      "description": "The directory where migrations are located.",
      "type": "string",
      "default": "migrations"
    },
    "logLevel": {
      "description": "The log level.",
      "type": "string",
      "default": "info",
      "enum": ["error", "warn", "info", "debug"]
    },
    "lifecycleHook": {
      "description": "The lifecycle hook to run.",
      "type": "string",
      "default": "before:deploy",
      "enum": ["before:deploy", "after:deploy"]
    },
    "from": {
      "description": "The namespace and version to rollback.",
      "type": "string"
    },
    "to": {
      "description": "The namespace and version to rollback to.",
      "type": "string"
    },
    "yes": {
      "description": "Skip confirmation prompt.",
      "type": "boolean",
      "default": false
    },
    "migrationTableName": {
      "description": "The name of the table to store the migration states.",
      "type": "string"
    }
  },
  "required": ["env", "cwd", "from"]
}
