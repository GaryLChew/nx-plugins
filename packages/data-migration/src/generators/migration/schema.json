{
  "$schema": "http://json-schema.org/schema",
  "$id": "Migration",
  "title": "",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project you'd like to add to the migration",
      "x-prompt": "What is the project you want to add to the migration?",
      "x-priority": "important"
    },
    "namespace": {
      "type": "string",
      "description": "The namespace of the migration, the namespace must be unique for the environment (e.g. users)",
      "x-prompt": "What is the migration namespace? (e.g. users)",
      "x-priority": "important"
    },
    "name": {
      "type": "string",
      "description": "The name of the migration, this will be used to generate the migration file name (e.g. create-table)",
      "x-prompt": "What is the migration name? (e.g. create-table)",
      "x-priority": "important"
    },
    "description": {
      "type": "string",
      "description": "The description of the migration (e.g. Create User Table)",
      "x-prompt": "What is the migration description? (e.g. Create User Table)"
    },
    "migrationsDir": {
      "type": "string",
      "description": "The directory where the migration files will be generated",
      "default": "src/migrations"
    },
    "migrationProvider": {
      "type": "string",
      "description": "The directory where the migration files will be generated",
      "default": "dynamodb",
      "enum": ["standard", "dynamodb"],
      "x-prompt": {
        "type": "list",
        "message": "What is the migration provider?",
        "items": ["dynamodb", "standard"]
      }
    },
    "lifecycleHook": {
      "type": "string",
      "description": "The lifecycle hook to run the migration",
      "default": "post",
      "enum": ["before:deploy", "after:deploy"],
      "x-prompt": {
        "type": "list",
        "message": "What is the lifecycle hook to run the migration?",
        "items": ["before:deploy", "after:deploy"]
      }
    },
    "parentVersion": {
      "type": "string",
      "description": "The parent version of the migration, only applicable for after:deploy migrations",
      "x-prompt": "What is the parent version of the migration, only applicable for after:deploy migrations?"
    },
    "addStream": {
      "type": "boolean",
      "description": "Add a stream to the migration, only available for dynamodb migrations (streams enable real-time data migration)",
      "default": false,
      "x-prompt": "Do you want to add a stream to the migration, only available for dynamodb migrations?"
    },
    "baseline": {
      "type": "boolean",
      "description": "Flag this migration as baseline",
      "default": false,
      "x-prompt": "Do you want to flag this migration as baseline?"
    }
  },
  "required": ["project", "name", "namespace"]
}
