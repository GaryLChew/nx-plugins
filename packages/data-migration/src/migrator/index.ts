export * from './types';
export * from './logger';
export * from './migration';
export * from './dynamodb';
