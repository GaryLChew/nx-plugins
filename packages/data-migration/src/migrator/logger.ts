import _ from 'lodash';
import chalk from 'chalk';

type ConsoleTypes = 'debug' | 'info' | 'warn' | 'error';

enum Levels {
  debug = 1,
  info = 2,
  warn = 3,
  error = 4,
}

export class CLILogger {
  private readonly logLevel: Levels = Levels.info;

  constructor(public readonly level: string) {
    const key = Object.entries(Levels).find(
      ([objectKey]) => objectKey === level,
    );
    if (!_.isNil(key) && key.length > 0) {
      this.logLevel = key[1] as Levels;
    }
  }

  public info(message: string): void {
    this.logEntry(Levels.info, message);
  }

  public error(message?: string): void {
    this.logEntry(Levels.error, message);
  }

  public warn(message: string): void {
    this.logEntry(Levels.warn, message);
  }

  public debug(message: string): void {
    this.logEntry(Levels.debug, message);
  }

  private logEntry(level: Levels, message: string): void {
    if (this.logLevel <= level) {
      const levelStr = Levels[level] as ConsoleTypes;
      const currentTime = new Date().toISOString();
      const color = this.getLevelColor(level);
      const lv = color(`[${levelStr}]`.padEnd(7, ' '));

      const prefix = chalk.bold(`[${currentTime}]:${lv}`);
      console[levelStr](`${prefix} ${message}`);
    }
  }

  private getLevelColor(level: Levels) {
    switch (level) {
      case Levels.debug:
        return chalk.gray;
      case Levels.info:
        return chalk.blue;
      case Levels.warn:
        return chalk.yellow;
      case Levels.error:
        return chalk.red;
    }
  }
}
