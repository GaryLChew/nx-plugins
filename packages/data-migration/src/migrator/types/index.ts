export * from './state';
export * from '../migration';
export * from './logger';
export * from './lifecycle-hook';
