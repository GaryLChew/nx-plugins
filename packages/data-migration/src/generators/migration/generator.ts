import { toKebabCase } from '@nxlv/util';
import {
  updateProjectConfiguration,
  formatFiles,
  generateFiles,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import path from 'path';
import { MigrationGeneratorSchema } from './schema';
import { globSync } from 'glob';
import { LifecycleHook } from '../../migrator';

function normalizeOptions(options: MigrationGeneratorSchema) {
  return {
    ...options,
    description: options.description || '',
    name: toKebabCase(options.name),
  };
}

function addFiles(
  tree: Tree,
  projectRoot: string,
  options: MigrationGeneratorSchema,
) {
  const date = new Date();
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');
  const formattedDate = `${year}${month}${day}`;
  const { filename, version } = getMigrationFilename(
    tree,
    projectRoot,
    options,
    formattedDate,
  );

  const templateOptions = {
    ...options,
    filename,
    ext: 'ts',
    version,
    lifecycleHook:
      options.lifecycleHook === LifecycleHook.BEFORE_DEPLOY
        ? 'BEFORE_DEPLOY'
        : 'AFTER_DEPLOY',
    template: '',
    dot: '.',
  };

  generateFiles(
    tree,
    path.join(__dirname, 'files', options.migrationProvider),
    path.join(
      projectRoot,
      options.migrationsDir,
      options.namespace,
      options.lifecycleHook === LifecycleHook.AFTER_DEPLOY ? 'post-deploy' : '',
    ),
    templateOptions,
  );

  if (options.addStream) {
    generateFiles(
      tree,
      path.join(__dirname, 'files', 'stream'),
      path.join(
        projectRoot,
        options.migrationsDir,
        options.namespace,
        options.lifecycleHook === LifecycleHook.AFTER_DEPLOY
          ? 'post-deploy'
          : '',
      ),
      templateOptions,
    );
  }
}

function getMigrationFilename(
  tree: Tree,
  projectRoot: string,
  options: MigrationGeneratorSchema,
  formattedDate: string,
  versionSeq = 1,
) {
  const version = parseInt(`${formattedDate}${versionSeq}`);
  const filename = `${formattedDate}${versionSeq}-${options.name}`;
  const baseFolder = path.join(
    projectRoot,
    options.migrationsDir,
    options.namespace,
  );

  const exists =
    globSync(path.join(baseFolder, `**/${version}-*.ts`)).length > 0;
  if (exists) {
    return getMigrationFilename(
      tree,
      projectRoot,
      options,
      formattedDate,
      versionSeq + 1,
    );
  }

  return { filename, version };
}

export default async function (tree: Tree, options: MigrationGeneratorSchema) {
  const normalizedOptions = normalizeOptions(options);

  const projectConfig = readProjectConfiguration(
    tree,
    normalizedOptions.project,
  );

  if ('migrate' in projectConfig.targets === false) {
    projectConfig.targets.migrate = {
      executor: '@nxlv/data-migration:migrate',
      options: {
        cwd: projectConfig.root,
        migrationsDir: normalizedOptions.migrationsDir,
      },
    };

    updateProjectConfiguration(tree, normalizedOptions.project, projectConfig);
  }
  if ('migrate-rollback' in projectConfig.targets === false) {
    projectConfig.targets['migrate-rollback'] = {
      executor: '@nxlv/data-migration:rollback',
      options: {
        cwd: projectConfig.root,
        migrationsDir: normalizedOptions.migrationsDir,
      },
    };

    updateProjectConfiguration(tree, normalizedOptions.project, projectConfig);
  }

  normalizedOptions.migrationsDir =
    projectConfig.targets.migrate.options.migrationsDir;

  addFiles(tree, projectConfig.root, normalizedOptions);

  await formatFiles(tree);
}
