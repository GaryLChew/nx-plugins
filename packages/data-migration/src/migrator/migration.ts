import { NetworkMode, AssignPublicIp } from '@aws-sdk/client-ecs';
import { CLILogger } from './logger';
import { LifecycleHook } from './types';

export type RemoteConfigParam = {
  value: string;
  type?: 'plain' | 'ssm';
  decrypt?: boolean;
};

export type RemoteConfig = {
  type: 'ecs';
  config: RemoteEcsConfig;
};

export type RemoteEcsConfig = {
  cluster: RemoteConfigParam;
  executionRoleArn: RemoteConfigParam;
  taskRoleArn: RemoteConfigParam;
  cpu: number;
  memory: number;
  networkMode: NetworkMode;
  subnetIds: RemoteConfigParam;
  securityGroupId: RemoteConfigParam;
  assignPublicIp?: AssignPublicIp;
};

export type MigrationOptions = {
  namespace: string;
  version: number;
  name: string;
  description?: string;
  lifecycleHook?: LifecycleHook;
  parentVersion?: number;
  baseline?: boolean;
  remote?: RemoteConfig;
};

export function Migration(options: MigrationOptions) {
  return function <T extends { new (...args) }>(constructor: T) {
    return class extends constructor {
      namespace = options.namespace;
      version = options.version;
      name = options.name;
      description = options.description;
      lifecycleHook = options.lifecycleHook ?? LifecycleHook.BEFORE_DEPLOY;
      parentVersion = options.parentVersion;
      baseline = options.baseline;
      remote = options.remote;
    };
  };
}

export abstract class MigrationBase {
  public readonly namespace: string;
  public readonly version: number;
  public readonly name: string;
  public readonly description: string;
  public readonly lifecycleHook: LifecycleHook;
  public readonly parentVersion?: number;
  public readonly baseline?: boolean;
  public readonly remote?: RemoteConfig;

  public distPath: string;
  public path: string;
  public logger: CLILogger;

  abstract up(): Promise<void>;
  abstract down(): Promise<void>;

  async sleep(seconds: number) {
    return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
  }

  async condition(): Promise<boolean> {
    return true;
  }
}
