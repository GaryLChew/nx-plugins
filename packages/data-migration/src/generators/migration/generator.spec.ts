import { vi } from 'vitest';

const globSync = vi.hoisted(() => vi.fn().mockReturnValue([]));
vi.mock('glob', () => ({ globSync }));

import { libraryGenerator } from '@nx/js';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import {
  Tree,
  readProjectConfiguration,
  updateProjectConfiguration,
} from '@nx/devkit';

import generator from './generator';
import { MigrationGeneratorSchema } from './schema';
import { LifecycleHook } from '../../migrator';

describe('migration generator', () => {
  let appTree: Tree;
  const options: MigrationGeneratorSchema = {
    name: 'migration-name',
    project: 'test',
    namespace: 'namespace',
    migrationsDir: 'src/migrations',
    addStream: false,
    lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
    migrationProvider: 'standard',
    parentVersion: undefined,
    baseline: false,
  };

  beforeEach(async () => {
    appTree = createTreeWithEmptyWorkspace({
      layout: 'apps-libs',
    });

    await libraryGenerator(appTree, {
      name: 'test',
      tags: 'scope:test,type:lib',
    });

    vi.useFakeTimers().setSystemTime(new Date('2023-01-01 12:00:00').getTime());
  });

  it('should run successfully', async () => {
    await generator(appTree, options);

    const projectConfig = readProjectConfiguration(appTree, 'test');
    expect(projectConfig.targets.migrate).toBeDefined();
    expect(projectConfig.targets.migrate.executor).toEqual(
      '@nxlv/data-migration:migrate',
    );
    expect(projectConfig.targets.migrate.options.migrationsDir).toEqual(
      'src/migrations',
    );

    expect(projectConfig.targets['migrate-rollback']).toBeDefined();
    expect(projectConfig.targets['migrate-rollback'].executor).toEqual(
      '@nxlv/data-migration:rollback',
    );
    expect(
      projectConfig.targets['migrate-rollback'].options.migrationsDir,
    ).toEqual('src/migrations');

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();

    globSync
      .mockReturnValueOnce(['202301011-migration-name.ts'])
      .mockReturnValueOnce([]);

    await generator(appTree, options);

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301012-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with custom migrations path', async () => {
    await generator(appTree, {
      ...options,
      migrationsDir: 'src/custom/migrations',
    });

    const projectConfig = readProjectConfiguration(appTree, 'test');
    expect(projectConfig.targets.migrate.options.migrationsDir).toEqual(
      'src/custom/migrations',
    );

    expect(
      appTree.read(
        'libs/test/src/custom/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with after deploy lifecycle', async () => {
    await generator(appTree, {
      ...options,
      lifecycleHook: LifecycleHook.AFTER_DEPLOY,
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/post-deploy/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with after deploy lifecycle, dynamodb provider and stream', async () => {
    await generator(appTree, {
      ...options,
      migrationProvider: 'dynamodb',
      addStream: true,
      lifecycleHook: LifecycleHook.AFTER_DEPLOY,
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/post-deploy/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/post-deploy/202301011-migration-name.stream.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with after deploy lifecycle and parent version', async () => {
    await generator(appTree, {
      ...options,
      lifecycleHook: LifecycleHook.AFTER_DEPLOY,
      parentVersion: 202304031,
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/post-deploy/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with dynamodb provider', async () => {
    await generator(appTree, {
      ...options,
      migrationProvider: 'dynamodb',
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with dynamodb provider and stream', async () => {
    await generator(appTree, {
      ...options,
      migrationProvider: 'dynamodb',
      addStream: true,
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301011-migration-name.stream.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should not update the project when the migration already exists', async () => {
    const projectConfig = readProjectConfiguration(appTree, 'test');
    projectConfig.targets.migrate = {
      executor: '@nxlv/data-migration:migrate',
      options: {
        migrationsDir: 'src/custom/migrations',
      },
    };
    projectConfig.targets['migrate-rollback'] = {
      executor: '@nxlv/data-migration:rollback',
      options: {
        migrationsDir: 'src/custom/migrations',
      },
    };

    updateProjectConfiguration(appTree, 'test', projectConfig);

    await generator(appTree, options);

    expect(
      appTree.read(
        'libs/test/src/custom/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();

    globSync
      .mockReturnValueOnce(['202301011-migration-name.ts'])
      .mockReturnValueOnce([]);

    await generator(appTree, options);

    expect(
      appTree.read(
        'libs/test/src/custom/migrations/namespace/202301012-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });

  it('should run successfully with baseline flag', async () => {
    await generator(appTree, {
      ...options,
      baseline: true,
    });

    expect(
      appTree.read(
        'libs/test/src/migrations/namespace/202301011-migration-name.ts',
        'utf-8',
      ),
    ).toMatchSnapshot();
  });
});
