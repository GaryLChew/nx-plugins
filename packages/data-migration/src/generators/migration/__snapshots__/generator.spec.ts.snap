// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migration generator > should not update the project when the migration already exists 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should not update the project when the migration already exists 2`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301012,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully 2`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301012,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with after deploy lifecycle 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.AFTER_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with after deploy lifecycle and parent version 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.AFTER_DEPLOY,
  parentVersion: 202304031,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with after deploy lifecycle, dynamodb provider and stream 1`] = `
"import {
  Migration,
  DynamoDBMigrationBase,
  LifecycleHook,
} from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.AFTER_DEPLOY,
})
export default class extends DynamoDBMigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with after deploy lifecycle, dynamodb provider and stream 2`] = `
"import { DynamoDBStreamHandler } from 'aws-lambda';

export const handler: DynamoDBStreamHandler = async (event) => {
  console.debug('Event', JSON.stringify(event));

  // IMPLEMENT YOUR LOGIC HERE

  console.log('Parsed records', event.Records.length);
};
"
`;

exports[`migration generator > should run successfully with baseline flag 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
  baseline: true,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with custom migrations path 1`] = `
"import { Migration, MigrationBase, LifecycleHook } from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends MigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with dynamodb provider 1`] = `
"import {
  Migration,
  DynamoDBMigrationBase,
  LifecycleHook,
} from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends DynamoDBMigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with dynamodb provider and stream 1`] = `
"import {
  Migration,
  DynamoDBMigrationBase,
  LifecycleHook,
} from '@nxlv/data-migration';

@Migration({
  namespace: 'namespace',
  version: 202301011,
  name: 'migration-name',
  description: '',
  lifecycleHook: LifecycleHook.BEFORE_DEPLOY,
})
export default class extends DynamoDBMigrationBase {
  async up(): Promise<void> {
    this.logger.info(\`To be implemented\`);
  }

  async down(): Promise<void> {
    this.logger.warn('To be implemented');
  }
}
"
`;

exports[`migration generator > should run successfully with dynamodb provider and stream 2`] = `
"import { DynamoDBStreamHandler } from 'aws-lambda';

export const handler: DynamoDBStreamHandler = async (event) => {
  console.debug('Event', JSON.stringify(event));

  // IMPLEMENT YOUR LOGIC HERE

  console.log('Parsed records', event.Records.length);
};
"
`;
